src/ASM/LR/ASMu3D.C

```diff
@@ -374,40 +375,128 @@ void ASMu3D::closeFaces (int dir, int basis, int master)
-  std::vector<LR::Basisfunction*> thisEdge;
-  if (dir == -1)
-    lrspline->getEdgeFunctions(thisEdge, LR::WEST, 1);
-  else if (dir == 1)
-    lrspline->getEdgeFunctions(thisEdge, LR::EAST, 1);
-  else if (dir == -2)
-    lrspline->getEdgeFunctions(thisEdge, LR::SOUTH, 1);
-  else if (dir == 2)
-    lrspline->getEdgeFunctions(thisEdge, LR::NORTH, 1);
-  else if (dir == -3)
-    lrspline->getEdgeFunctions(thisEdge, LR::BOTTOM, 1);
-  else if (dir == 3)
-    lrspline->getEdgeFunctions(thisEdge, LR::TOP, 1);
-
-  std::cout << "\nNumber of constraints: " << thisEdge.size() << std::endl;
+  // fetch the right basis to consider
+  LR::LRSplineVolume* lr = this->getBasis(basis);
 
-  // enforce the boundary conditions
-  for(LR::Basisfunction* b  : thisEdge)
-    this->prescribe(b->getId()+1,dof,bcode);
+  // get all elements and functions on this edge
+  std::vector<LR::Basisfunction*> edgeFunctions;
+  std::vector<LR::Element*>       edgeElements;
+  lr->getEdgeFunctions(edgeFunctions,edge);
+  lr->getEdgeElements (edgeElements ,edge);
+
+  // build up the local element/node correspondence needed by the projection
+  // call on this edge by ASMu2D::updateDirichlet()
+  DirichletFace de(edgeFunctions.size(), edgeElements.size(), dof, code, basis);
+  de.edg  = edge;
+  de.lr   = lr;
+
+  // find the corners since these are not to be included in the L2-fitting
+  // of the inhomogenuous dirichlet boundaries; corners are interpolatory.
+  // Optimization note: loop over the "edge"-container to manually pick up
+  // the end nodes. LRspine::getEdgeFunctions() does a global search.
+  std::vector<LR::Basisfunction*> c1, c2;
+  static const std::map<LR::parameterEdge,std::array<std::array<int,3>,4>> corners = {{
+        { LR::WEST, {{{-1, -1, -1},
+                      {-1,  1, -1},
+                      {-1, -1,  1},
+                      {-1,  1,  1}}}},
+        { LR::EAST,{{{  1, -1, -1},
+                      { 1,  1, -1},
+                      { 1, -1,  1},
+                      { 1,  1,  1}}}},
+        {LR::SOUTH,{{{-1, -1, -1},
+                      { 1, -1, -1},
+                      {-1, -1,  1},
+                      { 1, -1,  1}}}},
+        {LR::NORTH,{{{-1, -1, -1},
+                      { 1, -1, -1},
+                      {-1, -1,  1},
+                      { 1, -1,  1}}}},
+       {LR::BOTTOM,{{{-1, -1, -1},
+                      { 1, -1, -1},
+                      {-1,  1, -1},
+                      { 1,  1, -1}}}},
+          {LR::TOP,{{{-1, -1,  1},
+                      { 1, -1,  1},
+                      {-1,  1,  1},
+                      { 1,  1,  1}}}},
+  }};
+
+  int* c = de.corners;
+  for (const auto& it : corners.find(edge)->second)
+    *c++ = this->getCorner(it[0], it[1], it[2], basis);
+
+  int j = 0;
+  int bcode = abs(code);
+  for (auto b : edgeFunctions)
+  {
+    de.MLGN[j++] = b->getId();
+    int* cid = std::find(de.corners, de.corners+4, b->getId()+ofs);
```

Comment:
Can't see why you need de.MLGN to be a `map<int,int>` here. Think you can achieve the same with an `IntVec` (as the MLGE member is) and just use `push_back` to build it here. Is the map-functionality used somewhere I did not see it? If not, consider revising. The same also in ASMu2D also btw (probably a copy-paste from there).